### ISSUE TITLE
Extend `class CanBus` with `threading`

### ISSUE THREAD

**Issue by WK-GiHu** [JohannesAppel/johannesappel-#7]
@WK-GiHu 
I'm not familiar with the term and implementation. 
What is necessary to import? 
Can I create a new module to test this out rather?

_Originally posted by @JohannesAppel in https://github.com/JohannesAppel/johannesappel-/issues/2#issuecomment-567459973_

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>Can I create a new module to test this out rather?

Don't create a new module, backup your current version instead.

Extend `canbus.py` with:
```
import threading, queue

class CanBus(threading.Thread):
    def __init__(self, ....
        super().__init__()
        ...

        self.queue = queue.Queue()
        self._is_alive = threading.Event()
        self._is_alive.set()

    def get(self):
        # None blocking get
        try:
           data = self.queue.get(False)
        except queue.Empty:
           return None
        return data
        
    def stop(self):
        print('CanBus.stop()')
        self._is_alive.clear()
        while not self.queue.empty():
            self.queue.get()
    

    def run(self):
        # copy my sceleton
```

Move from `__main__` into a new function `def read4(cbus):` all Lines 61 to 75.

New in `__main__`:
```
cbus.start()

# For testing we need to terminate the `Thread`
max_frames = 100
for _ in range(max_frames):
    frame = cbus.get()
    print(frame)

cbus.stop()
print('EXIT')
```

**Comment by JohannesAppel** [comment]
@WK-GiHu 
>` except ???.QueueEmpty:`

What is the `???` ?

> def run(self):
        # copy my sceleton

What should I copy there? I'm a bit confused about that part.

I hope I moved everything correctly, it's a bit hard to follow.

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>What is the ??? ?

Didn't recall the correct statment at once, see my updated comment: https://github.com/JohannesAppel/johannesappel-/issues/7#issuecomment-567471798

---

>What should I copy there?

See: https://github.com/JohannesAppel/johannesappel-/issues/2#issuecomment-567450902

----

>I hope I moved everything correctly

You will get `Traceback` if not.
BTW: Which IDE, if any, you are using and which Python version?

**Comment by JohannesAppel** [comment]
@WK-GiHu 
> BTW: Which IDE, if any, you are using and which Python version?

I'm working through eclipse as my IDE, and it's python 3.7 if I remember correctly. 

> See: #2 (comment)

I only copied the producer part. 
It had an error I think. There was a `.` missing after the `self` in 
>self_is_alive.clear()

**Comment by JohannesAppel** [comment]
@WK-GiHu 
The program output is as follows:
```
canlib version: 8.30
Using channel: Kvaser Leaf Professional HS (channel 0), EAN: 73-30130-00404-7
None
 . . .
None
None
Traceback (most recent call last):
  File "C:\Users\Hex\eclipse-workspace\DataLogger\src\canbus.py", line 121, in <module>
None
None
None
None
CanBus.stop()
    cbus.stop()
  File "C:\Users\Hex\eclipse-workspace\DataLogger\src\canbus.py", line 38, in stop
    while not self.queue.is_empty():
AttributeError: 'Queue' object has no attribute 'is_empty'
```

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>AttributeError: 'Queue' object has no attribute 'is_empty'

See my updated comment: https://github.com/JohannesAppel/johannesappel-/issues/7#issuecomment-567471798, should read: `self.queue.empty()`

---

Nothing in the Queue?
We need a delay to give the `Thread` a timeslice, add before Line
https://github.com/JohannesAppel/johannesappel-/blob/b13c1c9ba64febf6d851433613677daa517b6399/DataLogger/src/canbus.py#L119
```
   # Import statement at the top of `canbus.py`
   import time
   ...
   time.sleep(0.1)
```


Add a trace print before Line
https://github.com/JohannesAppel/johannesappel-/blob/b13c1c9ba64febf6d851433613677daa517b6399/DataLogger/src/canbus.py#L56
```
    print('\tqueue.put({})'.format(item))
```

**Comment by JohannesAppel** [comment]
@WK-GiHu 
The output is still a bunch of `None`'s 
But it doesn't give `Traceback` errors anymore.

I does have the following bottom line at the end of the output:
```
CanBus.stop()
EXIT
	queue.put(Frame(id=2, data=bytearray(b'\xa4\x0b\x06\x00\x04\x00\x04\x00'), dlc=8, flags=<MessageFlag.STD: 2>, timestamp=10116))
```

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>It had an error I think. There was a . missing after the self in
>>    self_is_alive.clear()

Yes, replace it with:
```
       except:  # Break at unknow error
          self.stop()
          continue
```
---

>The output is still a bunch of None's

Get rid of the `None`, add before Line
https://github.com/JohannesAppel/johannesappel-/blob/b13c1c9ba64febf6d851433613677daa517b6399/DataLogger/src/canbus.py#L120
```
    if frame is None:
        none_count +=1
    else:
        print(frame)

    ...
    print('EXIT, none_count:{}'.format())
```
and before Line
https://github.com/JohannesAppel/johannesappel-/blob/b13c1c9ba64febf6d851433613677daa517b6399/DataLogger/src/canbus.py#L117
```
none_count = 0
```

>I does have the following bottom line at the end of the output:

Only **one**?

### Correction:
https://github.com/JohannesAppel/johannesappel-/blob/b13c1c9ba64febf6d851433613677daa517b6399/DataLogger/src/canbus.py#L46

should be `self.ch.read()`

**Comment by JohannesAppel** [comment]
@WK-GiHu 
>Only one?

Yes, there was only one line.

I updated the code.

**Comment by JohannesAppel** [comment]
@WK-GiHu 
Current output:
```
canlib version: 8.30
Using channel: Kvaser Leaf Professional HS (channel 0), EAN: 73-30130-00404-7
CanBus.stop()
	queue.put(Frame(id=2, data=bytearray(b'\xa4\x0b\x06\x00\x04\x00\x04\x00'), dlc=8, flags=<MessageFlag.STD: 2>, timestamp=10030))
Traceback (most recent call last):
  File "C:\Users\Hex\eclipse-workspace\DataLogger\src\canbus.py", line 130, in <module>
    print('EXIT, none_count:{}'.format())
IndexError: tuple index out of range
```

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>`IndexError: tuple index out of range`

Correct, should be: `print('EXIT, none_count:{}'.format(none_count))

**Comment by JohannesAppel** [comment]
@WK-GiHu 
Newest output:
```
canlib version: 8.30
Using channel: Kvaser Leaf Professional HS (channel 0), EAN: 73-30130-00404-7
CanBus.stop()
EXIT, none_count:100
```

**Comment by WK-GiHu** [comment]
@JohannesAppel 

Extend the trace doing:
```
    def run(self):
        while self._is_alive.is_set():
            item = None
            try:
                item = self.ch.read()
            except (canlib.canNoMsg):
                self.queue.put('except canlib.canNoMsg')
                pass
            except Exception as e:  # Break at unknow error
                self.queue.put(e)
                self.stop()
                continue
    
        if item is not None:
            # Possible to decode data before
            # item.data = text(item.data)
            print('\tqueue.put({})'.format(item))
            self.queue.put(item)
        else:
            self.queue.put('item is None')

```

---

In the meantime i read at **KVASER**: [Handles are thread-specific](https://www.kvaser.com/canlib-webhelp/page_user_guide_threads.html), i think about it.

**Comment by WK-GiHu** [comment]
@JohannesAppel 

## GOT IT

This block should inside the `while`, indet it
https://github.com/JohannesAppel/johannesappel-/blob/fe4b989da3929f3c8d45200e703a237f58c1784a/DataLogger/src/canbus.py#L54

**Comment by JohannesAppel** [comment]
@WK-GiHu 
Should I remove the code from your previous comment?

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>Should I remove the code from your previous comment?

Do **one** run with it, may be interesting. Then disable it with `# ...` maybe for later debuging.

**Comment by JohannesAppel** [comment]
@WK-GiHu 
Output as the code is now included with previous comments code:
```
canlib version: 8.30
Using channel: Kvaser Leaf Professional HS (channel 0), EAN: 73-30130-00404-7
	queue.put(Frame(id=1, data=bytearray(b'\x00\x1e\x84\x80'), dlc=4, flags=<MessageFlag.STD: 2>, timestamp=24))
	queue.put(Frame(id=2, data=bytearray(b'\xa4\x0b\x06\x00\x04\x00\x04\x00'), dlc=8, flags=<MessageFlag.STD: 2>, timestamp=24))
	queue.put(Frame(id=3, data=bytearray(b'\x05\x00\x05\x00\x05\x00\x04\x00'), dlc=8, flags=<MessageFlag.STD: 2>, timestamp=24))
	queue.put(Frame(id=4, data=bytearray(b'\xbc\x00\t\x00'), dlc=4, flags=<MessageFlag.STD: 2>, timestamp=24))
 . . .
	queue.put(Frame(id=2, data=bytearray(b'\xa4\x0b\x06\x00\xb5\x0b\x04\x00'), dlc=8, flags=<MessageFlag.STD: 2>, timestamp=10021))
	queue.put(Frame(id=3, data=bytearray(b'\x07\x00\x05\x00\x04\x00\x05\x00'), dlc=8, flags=<MessageFlag.STD: 2>, timestamp=10021))
	queue.put(Frame(id=4, data=bytearray(b'\xbd\x00\r\x00'), dlc=4, flags=<MessageFlag.STD: 2>, timestamp=10021))
item is None
CanBus.stop()
EXIT, none_count:0
```

**Comment by JohannesAppel** [comment]
@WK-GiHu 
I've run the code a few times now. Each output differs a bit.
Somewhere in the output the following line occurs randomly:
```
Frame(id=3, data=bytearray(b'\x05\x00\x05\x00\x05\x00\x05\x00'), dlc=8, flags=<MessageFlag.STD: 2>, timestamp=230)
```

And sometimes this at the end also
```
queue.put(Frame(id=1, data=bytearray(b'\x00\x1e\x84\x80'), dlc=4, flags=<MessageFlag.STD: 2>, timestamp=10955))EXIT, none_count:0
```

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>Somewhere in the output the following line occurs randomly:

You mean the sequence `id` = `1, 2, 3, 4` is brocken?

The `Queue` is a `FIFO`, means First In First Out.

---

>`item is None`

It seems you get at least **one** `canlib.canNoMsg` from `self.ch.read()`.
This could indicate a timeout on the **CAN BUS**, think about counting these `canlib.canNoMsg` `.read()` and get a report later on how much you get.

**Comment by JohannesAppel** [comment]
@WK-GiHu 
It doesn't look like the sequence is broken, since the output stays in chronological order. 

We never call the `read4(cbus)` function?

What is the next step after this? 
Tomorrow will be my last day working on this for this year, going on vacation. But will be back next year. I appreciate the help so far, wouldn't have done it without you!

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>It doesn't look like the sequence is broken, since the output stays in chronological order.

>***the following line occurs randomly:***?

This is as you have running 2 `Thread` which both do `print(...`.
It's not guaranteed, that one `print(...` gets finished while the other `print(...` starts.
This happens randomly.


Disable Line
https://github.com/JohannesAppel/johannesappel-/blob/e0cc9cb310fbdd7a7332cff345db95b39413d5ed/DataLogger/src/canbus.py#L59

---

>We never call the read4(cbus) function?

It's not needed in this **Producer/Consumer** model.
As long as you are not satisfied with the current output, hold it.

---

>What is the next step after this?

1. Verify if the 10 sec delay occures:
  see: https://github.com/JohannesAppel/johannesappel-/issues/5#issuecomment-567542918

2. Implement the **Consumer** part
  see: https://github.com/JohannesAppel/johannesappel-/issues/8#issue-540405472

---

### Improvement
Add before Line
https://github.com/JohannesAppel/johannesappel-/blob/e0cc9cb310fbdd7a7332cff345db95b39413d5ed/DataLogger/src/canbus.py#L39
```
time.sleep(0.1)
```

**Comment by JohannesAppel** [comment]
@WK-GiHu 
This exception error output still occurs randomly throughout the output, but always after a full data pack (1 to 4) has been read. 
```
except canlib.canNoMsg
item is None
```
But I don't think it's really a problem. Will just `#` the prints out when the code is fully functional.

And then the timing is wrong. It's way more off than last time. I don't even see the values update now anymore. Previously there were only a delay, but now it's as if it never comes to the data point where I flipped the switch. 
The timing is only wrong on the canbus module since the main is working yet.

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>This exception error output still occurs randomly

As i wrote, https://github.com/JohannesAppel/johannesappel-/issues/7#issuecomment-567517665, this is HARDWARE related. You do `ch.read()` faster then the hardware could/does deliver.

>And then the timing is wrong.
Yes, we use 
https://github.com/JohannesAppel/johannesappel-/blob/ae76b9efcba3d2561eb5c23a09d59f964b4c1f30/DataLogger/src/canbus.py#L128

decrease it to `.sleep(0.01)`

**WARNING**: If the **Producer** `queue.put(...)` is much more than the **Consumer** `queue.get()` then the `Queue` will overload and get stuck. 
Raises exception [`queue.Full`](https://docs.python.org/3/library/queue.html#queue.Full)

To verify, add:
```
class CanBus...
    def __init__(...
        ...
        self.frame_count = 0

    def run(...
        ...
        try:
            item = self.ch.read()
            self.fram_count += 1
        except (canlib.canNoMsg):
           ...
```
On **EXIT** print `cbus.frame_count` and compare with `max_frames`.

I get the feeling, doing `tkinter insert(` with a intervall of `10 msec` would overload `tkinter` also?

**Comment by JohannesAppel** [comment]
@WK-GiHu 
`tkinter` outputs data now after the changes, but the tkinter app doesn't update the vlaues a of yet. It just appends the values at the bottom of the previous print of values.
